File: plugins/java/src/main/java/org/bu1ld/plugins/java/Protocol.java
package org.bu1ld.plugins.java;

import com.google.gson.annotations.SerializedName;
import java.util.List;
import java.util.Map;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class Protocol {
    public static record Metadata(
        String id,
        String namespace,
        List<RuleSchema> rules,
        @SerializedName("config_fields")
        List<FieldSchema> configFields,
        @SerializedName("auto_configure")
        boolean autoConfigure
    ) {
    }

    public static record MetadataResult(Metadata metadata) {
    }

    public static record FieldSchema(String name, String type, boolean required) {
    }

    public static record RuleSchema(String name, List<FieldSchema> fields) {
    }

    public static record ExpandParams(Invocation invocation) {
    }

    public static record ExpandResult(List<TaskSpec> tasks) {
    }

    public static record PluginConfig(String namespace, Map<String, Object> fields) {
    }

    public static record ConfigureParams(PluginConfig config) {
    }

    public static record ConfigureResult(List<TaskSpec> tasks) {
    }

    public static record ExecuteParams(ExecuteRequest request) {
    }

    public static record ExecuteRequest(String namespace, String action, @SerializedName("work_dir") String workDir, Map<String, Object> params) {
    }

    public static record ExecuteResult(String output) {
    }

    public static record TaskSpec(
        String name,
        List<String> deps,
        List<String> inputs,
        List<String> outputs,
        List<String> command,
        TaskAction action
    ) {
    }

    public static record TaskAction(String kind, Map<String, Object> params) {
    }
}

File: plugins/typescript/src/protocol.ts
export const protocolVersion = 1;
export const pluginExecActionKind = "plugin.exec";

export type FieldType = "string" | "list" | "object" | "bool";

export interface Metadata {
  id: string;
  namespace: string;
  protocol_version: number;
  capabilities: string[];
  rules: RuleSchema[];
  config_fields?: FieldSchema[];
  auto_configure?: boolean;
}

export interface MetadataResult {
  metadata: Metadata;
}

export interface FieldSchema {
  name: string;
  type: FieldType;
  required?: boolean;
}

export interface RuleSchema {
  name: string;
  fields: FieldSchema[];
}

export interface Invocation {
  namespace: string;
  rule: string;
  target: string;
  fields?: Record<string, unknown>;
}

export interface ExpandParams {
  invocation: Invocation;
}

export interface ExpandResult {
  tasks: TaskSpec[];
}

export interface PluginConfig {
  namespace: string;
  fields?: Record<string, unknown>;
}

export interface ConfigureParams {
  config: PluginConfig;
}

export interface ConfigureResult {
  tasks: TaskSpec[];
}

export interface ExecuteRequest {
  namespace: string;
  action: string;
  work_dir: string;
  params?: Record<string, unknown>;
}

export interface ExecuteParams {
  request: ExecuteRequest;
}

export interface ExecuteResult {
  output?: string;
}

export interface TaskSpec {
  name: string;
  deps?: string[];
  inputs?: string[];
  outputs?: string[];
  command?: string[];
  action?: TaskAction;
}

export interface TaskAction {
  kind: string;
  params: Record<string, unknown>;
}

export function field(name: string, type: FieldType, required = false): FieldSchema {
  return { name, type, required: required || undefined };
}

export function rule(name: string, fields: FieldSchema[]): RuleSchema {
  return { name, fields };
}
